openprescribing/frontend/views/views.py

```diff
@@ -1062,6 +983,171 @@ def _financial_ytd_total(monthly_totals):
+                pk__in=request.POST.getlist("ncso_concessions_bookmarks")
+            )
+
+        # QuerySet.delete() returns a tuple whose first element is the number
+        # of records deleted.
+        count = (
+            org_bookmarks.delete()[0]
+            + search_bookmarks.delete()[0]
+            + ncso_concessions_bookmarks.delete()[0]
+        )
+
+        if count > 0:
+            msg = "Unsubscribed from %s alert" % count
+            if count > 1:
+                msg += "s"
+            messages.success(request, msg)
+
+        return redirect(reverse("bookmarks", args=[key]))
+
+    form = BookmarkListForm(
+        org_bookmarks=org_bookmarks,
+        search_bookmarks=search_bookmarks,
+        ncso_concessions_bookmarks=ncso_concessions_bookmarks,
+    )
+    count = (
+        search_bookmarks.count()
+        + org_bookmarks.count()
+        + ncso_concessions_bookmarks.count()
+    )
+
+    if count == 1:
+        single_bookmark = (
+            search_bookmarks.first()
+            or org_bookmarks.first()
+            or ncso_concessions_bookmarks.first()
+        )
+    else:
+        single_bookmark = None
+
+    ctx = {
+        "search_bookmarks": search_bookmarks,
+        "org_bookmarks": org_bookmarks,
+        "ncso_concessions_bookmarks": ncso_concessions_bookmarks,
+        "form": form,
+        "count": count,
+        "single_bookmark": single_bookmark,
+    }
+
+    return render(request, "bookmarks/bookmark_list.html", ctx)
+
+
+BOOKMARK_CLS_TO_FORM_CLS = {
+    OrgBookmark: OrgBookmarkForm,
+    SearchBookmark: SearchBookmarkForm,
+    NCSOConcessionBookmark: OrgBookmarkForm,
+}
+
+
+def _build_bookmark_form(bookmark_cls, initial):
+    """Build form for alert signup."""
+
+    form_cls = BOOKMARK_CLS_TO_FORM_CLS[bookmark_cls]
+    return form_cls(initial=initial)
+
+
+def _handle_bookmark_post(request, bookmark_cls):
+    """Create a bookmark, email the user, add confirmation message, and
+    redirect to bookmark's dashboard URL.
+    """
+
+    bookmark = _get_or_create_bookmark(request, bookmark_cls)
+    _send_alert_signup_confirmation(bookmark.user)
+    _add_confirmation_message(request, bookmark)
+    return redirect(bookmark.dashboard_url())
+
+
+def _get_or_create_bookmark(request, bookmark_cls):
+    """Get or create bookmark object.
+
+    Note that this will raise a ValidationError if the email address is invalid
+    (which shouldn't happen because it should be validated by the browser) or
+    or an IntegrityError if the submitted pct_id/practice_id/pcn_id doesn't
+    correspond to an existing PCT/Practice/PCN (which shouldn't happen because
+    the entity id should be set in the initial form by _build_bookmark_form().)
+    """
+
+    form_cls = BOOKMARK_CLS_TO_FORM_CLS[bookmark_cls]
+    form = form_cls(request.POST)
+    form.full_clean()
+    email = form.cleaned_data["email"]
+    user, _ = User.objects.get_or_create(username=email, defaults={"email": email})
```

Comment:
I think we should either use a case-insensitive match here or, more simply, force the email to lowercase before doing anything else with it. Either way, it would be better not to end up with duplicate users because of inconsistencies in email case.

Reply:
Good idea.

We have 37 users whose email address is not lowercase.  Should we convert those to lowercase?  And if so, should we use a data migration or do it with a one-off script?

Reply:
I think we should convert the existing users to lowercase using a one off script. Doing it with a data migration is just extra work for no real benefit.

Reply:
Any reason not to do this:

```
from django.contrib.auth.models import User

for user in User.objects.all():
    user.username = user.username.lower()
    user.email = user.email.lower()
    user.save()
```

?

Reply:
Not that I can think of!